File: detection/semgrep/fixtures/xff-for-logging.js
// Real-capture fixture for rule: mcp-authz-scope-from-request-header
//
// From audit: SurfSense and other real MCP servers read X-Forwarded-For for
// logging/analytics ONLY, not for authorization. The hardened rule must NOT
// fire on this pattern — it's the SILENT half of the two-way canary.
//
// SILENT: X-Forwarded-For read ONLY for logging, no authz decision.

async function logRequestForAnalytics(_args, extra) {
  // ok: mcp-authz-scope-from-request-header
  const clientIp = extra?.requestInfo?.headers?.["x-forwarded-for"];
  logger.info({ clientIp, event: "tool_call", timestamp: Date.now() });
}

async function logRequestWithXRealIp(_args, extra) {
  // ok: mcp-authz-scope-from-request-header
  const clientIp = extra?.requestInfo?.headers?.["x-real-ip"];
  logger.info({ clientIp, event: "tool_call" });
}

async function logRequestWithXClientIp(_args, extra) {
  // ok: mcp-authz-scope-from-request-header
  const clientIp = extra?.requestInfo?.headers?.["x-client-ip"];
  logger.info({ clientIp, event: "tool_call" });
}

File: detection/semgrep/fixtures/xff-for-logging.py
# Real-capture fixture for rule: mcp-authz-scope-from-request-header-py
#
# From audit: real FastMCP servers read X-Forwarded-For for logging/analytics
# ONLY, not for authorization. The hardened rule must NOT fire on this pattern
# — it's the SILENT half of the two-way canary.
#
# SILENT: X-Forwarded-For read ONLY for logging, no authz decision.

from fastmcp.server.dependencies import get_http_headers


def log_request_for_analytics():
    # ok: mcp-authz-scope-from-request-header-py
    headers = get_http_headers()
    client_ip = headers.get("x-forwarded-for")
    logger.info("tool call", extra={"client_ip": client_ip, "event": "tool_call"})


def log_request_with_x_real_ip():
    # ok: mcp-authz-scope-from-request-header-py
    headers = get_http_headers()
    client_ip = headers.get("x-real-ip")
    logger.info("tool call", extra={"client_ip": client_ip, "event": "tool_call"})


def log_request_with_x_client_ip():
    # ok: mcp-authz-scope-from-request-header-py
    headers = get_http_headers()
    client_ip = headers.get("x-client-ip")
    logger.info("tool call", extra={"client_ip": client_ip, "event": "tool_call"})